In [1]:
import pandas as pd
import nltk
from nltk.probability import FreqDist
from nltk.corpus import stopwords, words
from nltk import word_tokenize
import re
import numpy as np
from wordcloud import WordCloud, STOPWORDS
import matplotlib.pyplot as plt
from PIL import Image
import requests
from io import BytesIO

In [2]:
#Function to flatten 
flatten = lambda l: [item for sublist in l for item in sublist]

In [3]:
url = "https://raw.githubusercontent.com/LunaHub/Friends_social_data_analysis_2019/master/data/All_Friends_data.csv"
df = pd.read_csv(url, error_bad_lines=False)

In [4]:
df = df.drop("Unnamed: 0",axis=1)

In [5]:
mainch = ["joey","rachel","ross","chandler","monica","phoebe"]

In [6]:
# function for clearing lines
def cleanLines(line):
    #Exclude links
    line = re.sub(r'\([^)]*\)','',line)
    #Remove everything besides letters
    line = re.sub("([^a-z \t])|(\w+:\/\/\S+)", "", line)
    #Create a string for tweet
    line = [''.join(s for s in line)]
    return(line) 

In [7]:
#Find episodes  of each season
s1 = df[['Speaker','Text','Scene']][df['Episode'].str.match(pat = '(1[0-2][1-9])')].reset_index(drop=True)
s2 = df[['Speaker','Text','Scene']][df['Episode'].str.match(pat = '(2[0-2][1-9])')].reset_index(drop=True)
s3 = df[['Speaker','Text','Scene']][df['Episode'].str.match(pat = '(3[0-2][1-9])')].reset_index(drop=True)
s4 = df[['Speaker','Text','Scene']][df['Episode'].str.match(pat = '(4[0-2][1-9])')].reset_index(drop=True)
s5 = df[['Speaker','Text','Scene']][df['Episode'].str.match(pat = '(5[0-2][1-9])')].reset_index(drop=True)
s6 = df[['Speaker','Text','Scene']][df['Episode'].str.match(pat = '(6[0-2][1-9])')].reset_index(drop=True)
s7 = df[['Speaker','Text','Scene']][df['Episode'].str.match(pat = '(7[0-2][1-9])')].reset_index(drop=True)
s8 = df[['Speaker','Text','Scene']][df['Episode'].str.match(pat = '(8[0-2][1-9])')].reset_index(drop=True)
s9 = df[['Speaker','Text','Scene']][df['Episode'].str.match(pat = '(9[0-2][1-9])')].reset_index(drop=True)
s10 = df[['Speaker','Text','Scene']][df['Episode'].str.match(pat = '(10[0-2][1-9])')].reset_index(drop=True)

In [8]:
#Create list for what characters says in season 1. liness1[0] is what joey says, liness1[5] is phoebe
liness1 = []
for character in mainch:
    says = s1.Text[[character in i for i in s1.Speaker.str.split("\b")]].reset_index(drop=True)
    liness1.append(says)

liness2 = []
for character in mainch:
    says = s2.Text[[character in i for i in s2.Speaker.str.split("\b")]].reset_index(drop=True)
    liness2.append(says)
    
liness3 = []
for character in mainch:
    says = s3.Text[[character in i for i in s3.Speaker.str.split("\b")]].reset_index(drop=True)
    liness3.append(says)

liness4 = []
for character in mainch:
    says = s4.Text[[character in i for i in s4.Speaker.str.split("\b")]].reset_index(drop=True)
    liness4.append(says)

liness5 = []
for character in mainch:
    says = s5.Text[[character in i for i in s5.Speaker.str.split("\b")]].reset_index(drop=True)
    liness5.append(says)

liness6 = []
for character in mainch:
    says = s6.Text[[character in i for i in s6.Speaker.str.split("\b")]].reset_index(drop=True)
    liness6.append(says)
    
liness7 = []
for character in mainch:
    says = s7.Text[[character in i for i in s7.Speaker.str.split("\b")]].reset_index(drop=True)
    liness7.append(says)

liness8 = []
for character in mainch:
    says = s8.Text[[character in i for i in s8.Speaker.str.split("\b")]].reset_index(drop=True)
    liness8.append(says)
    
liness9 = []
for character in mainch:
    says = s9.Text[[character in i for i in s9.Speaker.str.split("\b")]].reset_index(drop=True)
    liness9.append(says)

liness10 = []
for character in mainch:
    says = s10.Text[[character in i for i in s10.Speaker.str.split("\b")]].reset_index(drop=True)
    liness10.append(says)
    

In [9]:
seasons = ["Season 1","Season 2","Season 3","Season 4","Season 5","Season 6","Season 7","Season 8","Season 9","Season 10","All in total"]

In [10]:
n = 0
joeylines = []
joeylines.append(liness1[n])
joeylines.append(liness2[n])
joeylines.append(liness3[n])
joeylines.append(liness4[n])
joeylines.append(liness5[n])
joeylines.append(liness6[n])
joeylines.append(liness7[n])
joeylines.append(liness8[n])
joeylines.append(liness9[n])
joeylines.append(liness10[n])


In [38]:
lines = ["joey","rachel","ross","chandler","monica","phoebe","gunther","jack","judy","geller","janice","mike","mom","dad"]
saystimes = []
for line in range(len(lines)):
    saysthis = []
    for n in joeylines:
        saysthis.append(sum(lines[line] in say for say in n))
    saysthis.append(sum(saysthis)) 
    saystimes.append(saysthis)
pd.DataFrame(saystimes, lines,seasons)

Season 1  Season 2  Season 3  Season 4  Season 5  Season 6  \
joey            18         9         5        26        27        30   
rachel          12         7        10        16        25        22   
ross            55        27        26        37        60        35   
chandler        32        37        38        53        45        62   
monica          18        12        17        15        35        23   
phoebe          23         8        11         8        22        13   
gunther          0         1         0         3         0         8   
jack             0         1         1         3         1         2   
judy             0         0         0         0         0         0   
geller           0         0         1         5         0         1   
janice           1         0         9         2         0         0   
mike            10         0         1         0         0         0   
mom              9         2         5         3         1         4   
dad              4         1         3         2         4         0   

          Season 7  Season 8  Season 9  Season 10  All in total  
joey            21        20        28         13           197  
rachel          29        42        23         17           203  
ross            35        46        44         33           398  
chandler        35        29        34         21           386  
monica          22        19        24         12           197  
phoebe          20        21         6         19           151  
gunther          0         0         3          0            15  
jack             4         1         3          1            17  
judy             0         0         0          0             0  
geller           0         0         0          2             9  
janice           0         0         0          0            12  
mike             0         1        12         11            35  
mom              2         6         6          4            42  
dad              1         1         0          3            19

In [39]:
n = 1
rachellines = []
rachellines.append(liness1[n])
rachellines.append(liness2[n])
rachellines.append(liness3[n])
rachellines.append(liness4[n])
rachellines.append(liness5[n])
rachellines.append(liness6[n])
rachellines.append(liness7[n])
rachellines.append(liness8[n])
rachellines.append(liness9[n])
rachellines.append(liness10[n])


In [40]:
saystimes = []
for line in range(len(lines)):
    saysthis = []
    for n in rachellines:
        saysthis.append(sum(lines[line] in say for say in n))
    saysthis.append(sum(saysthis)) 
    saystimes.append(saysthis)

pd.DataFrame(saystimes, lines,seasons)

Season 1  Season 2  Season 3  Season 4  Season 5  Season 6  \
joey            13         8         9        17        45        44   
rachel          15         6         1        17        17        18   
ross            86        37        92        50        72        89   
chandler        21        14        11        22        27        17   
monica          54        25        24        26        36        36   
phoebe          19        15        20        16        31        54   
gunther          1         2         3         2         1         2   
jack             4         0         1         0         4         1   
judy             0         0         0         0         0         0   
geller           2         1         2         1         1         3   
janice           1         0         3         0         1         0   
mike             1         0         0         0         0         0   
mom              8        12         1         6         2         2   
dad              6        10         9         2         1         7   

          Season 7  Season 8  Season 9  Season 10  All in total  
joey            92        71        47         42           388  
rachel          12        16        10          8           120  
ross            29        91        66         81           693  
chandler        23         9        11         11           166  
monica          51        16        22         16           306  
phoebe          53        24        31         20           283  
gunther          1         2         1          1            16  
jack             1         1         0          0            12  
judy             0         0         1          0             1  
geller           0         1         2          0            13  
janice           0         0         0          0             5  
mike             1         0         5          0             7  
mom              4         4        10          8            57  
dad              4         7         2          1            49

In [41]:
n = 2
rosslines = []
rosslines.append(liness1[n])
rosslines.append(liness2[n])
rosslines.append(liness3[n])
rosslines.append(liness4[n])
rosslines.append(liness5[n])
rosslines.append(liness6[n])
rosslines.append(liness7[n])
rosslines.append(liness8[n])
rosslines.append(liness9[n])
rosslines.append(liness10[n])

In [42]:
saystimes = []
for line in range(len(lines)):
    saysthis = []
    for n in rosslines:
        saysthis.append(sum(lines[line] in say for say in n))
    saysthis.append(sum(saysthis)) 
    saystimes.append(saysthis)

pd.DataFrame(saystimes, lines,seasons)

Season 1  Season 2  Season 3  Season 4  Season 5  Season 6  \
joey            31        19        25        27        30        21   
rachel          64        34        57        17        35        41   
ross            24        11        17        15        23        15   
chandler        37        14        36        25        22        24   
monica          29        19        19        19        15        24   
phoebe          19         8        11        13        23        23   
gunther          3         3         5         1         3         1   
jack             2         0         2         2         4         1   
judy             0         0         0         0         1         0   
geller           4         2         3         1         4         3   
janice           1         2         2         0         4         0   
mike             7         0         0         0         0         0   
mom             21         9         5         9         2         5   
dad              9        13         6         9         4         7   

          Season 7  Season 8  Season 9  Season 10  All in total  
joey            25        30        41         34           283  
rachel          17        67        46         51           429  
ross            15        25        29         12           186  
chandler        42         7        15         19           241  
monica          36        10        12         14           197  
phoebe          30         5        23         10           165  
gunther          2         1         1          3            23  
jack             0         1         1          1            14  
judy             0         0         0          0             1  
geller           3         5         5          1            31  
janice           0         2         0          0            11  
mike             0         0         9          5            21  
mom             10         8        14          3            86  
dad             19         6         7          4            84

In [43]:
n = 3
chandlerlines = []
chandlerlines.append(liness1[n])
chandlerlines.append(liness2[n])
chandlerlines.append(liness3[n])
chandlerlines.append(liness4[n])
chandlerlines.append(liness5[n])
chandlerlines.append(liness6[n])
chandlerlines.append(liness7[n])
chandlerlines.append(liness8[n])
chandlerlines.append(liness9[n])
chandlerlines.append(liness10[n])

In [44]:
saystimes = []
for line in range(len(lines)):
    saysthis = []
    for n in chandlerlines:
        saysthis.append(sum(lines[line] in say for say in n))
    saysthis.append(sum(saysthis)) 
    saystimes.append(saysthis)

pd.DataFrame(saystimes, lines,seasons)

Season 1  Season 2  Season 3  Season 4  Season 5  Season 6  \
joey            44        44        37        51        35        38   
rachel          20         5        15        11        12        18   
ross            41        12        28        27        28        32   
chandler        18        13         8        17         7        19   
monica          26        11        14         8        60        52   
phoebe          11         5         8        13        19        22   
gunther          0         1         1         2         0         0   
jack             0         3         0         1         0         3   
judy             0         0         0         1         0         1   
geller           1         1         0         0         2         2   
janice          14         6        21         5         0         0   
mike             3         0         0         0         0         0   
mom             16         2         6         3         4         4   
dad              4         1         1         0         0         6   

          Season 7  Season 8  Season 9  Season 10  All in total  
joey            34        15        32         27           357  
rachel          13        10        11          9           124  
ross            35        18        28         17           266  
chandler         6        11        19         10           128  
monica          85        23        51         39           369  
phoebe          11        13        12          5           119  
gunther          1         0         0          0             5  
jack             0         1         3          0            11  
judy             0         0         0          0             2  
geller           2         0         2          1            11  
janice           1         3         1          5            56  
mike             0         0         3          3             9  
mom              4         4         4          1            48  
dad             10         0         0          3            25

We can see that in season 5 a change happens - Joey stops being the most important person in Chandlers life and it then becomes Monica that he mentions the most

In [45]:
n = 4
monicalines = []
monicalines.append(liness1[n])
monicalines.append(liness2[n])
monicalines.append(liness3[n])
monicalines.append(liness4[n])
monicalines.append(liness5[n])
monicalines.append(liness6[n])
monicalines.append(liness7[n])
monicalines.append(liness8[n])
monicalines.append(liness9[n])
monicalines.append(liness10[n])

In [46]:
saystimes = []
for line in range(len(lines)):
    saysthis = []
    for n in monicalines:
        saysthis.append(sum(lines[line] in say for say in n))
    saysthis.append(sum(saysthis)) 
    saystimes.append(saysthis)

pd.DataFrame(saystimes, lines,seasons)

Season 1  Season 2  Season 3  Season 4  Season 5  Season 6  \
joey            43        15        32        27        32        38   
rachel          52        24        25        27        44        45   
ross            46        25        24        24        20        36   
chandler        37        10        20        21        61        77   
monica          27         5        10         6        13         7   
phoebe          46        14        38        27        34        37   
gunther          1         3         1         1         3         0   
jack             3         3         0         2         1         3   
judy             0         1         0         1         0         1   
geller           6         0         2         2         1         1   
janice           3         0         1         1         0         0   
mike             4         0         0         0         0         0   
mom              9        16         2         5         0         5   
dad              5        13         5         5         2         9   

          Season 7  Season 8  Season 9  Season 10  All in total  
joey            24        36        18         28           293  
rachel          21        23         8         13           282  
ross            37        20        17         15           264  
chandler        73        49        52         61           461  
monica          18        10        11          7           114  
phoebe          38        41        32         35           342  
gunther          0         0         0          1            10  
jack             2         0         3          1            18  
judy             1         0         0          0             4  
geller           4         1         4          2            23  
janice           3         1         0          2            11  
mike             0         0        14         10            28  
mom             11         3         3          0            54  
dad             16         3         9          2            69

In [47]:
n = 5
phoebelines = []
phoebelines.append(liness1[n])
phoebelines.append(liness2[n])
phoebelines.append(liness3[n])
phoebelines.append(liness4[n])
phoebelines.append(liness5[n])
phoebelines.append(liness6[n])
phoebelines.append(liness7[n])
phoebelines.append(liness8[n])
phoebelines.append(liness9[n])
phoebelines.append(liness10[n])

In [48]:
saystimes = []
for line in range(len(lines)):
    saysthis = []
    for n in phoebelines:
        saysthis.append(sum(lines[line] in say for say in n))
    saysthis.append(sum(saysthis))    
    saystimes.append(saysthis)

pd.DataFrame(saystimes, lines,seasons)

Season 1  Season 2  Season 3  Season 4  Season 5  Season 6  \
joey            28         8        16        18        35        22   
rachel          36         7        19        18        26        50   
ross            23        14        29        27        15        43   
chandler        24        19        19        15        32        24   
monica          37        15        41        18        24        22   
phoebe          16         6         9        20        11         9   
gunther          2         0         0         3         1         0   
jack             5         1         0         1         2         5   
judy             1         0         0         0         0         0   
geller           1         0         1         1         1         3   
janice           1         3         1         1         1         0   
mike            10         0         0         0         0         0   
mom             11         3        10        20         5         0   
dad              6        13         5         2         2         1   

          Season 7  Season 8  Season 9  Season 10  All in total  
joey            21        21        21         21           211  
rachel          35        21        33         23           268  
ross            29        14        37         17           248  
chandler        32        23         9         14           211  
monica          48        30        23         19           277  
phoebe          14        18        19          6           128  
gunther          0         2         0          2            10  
jack             0         0         2          1            17  
judy             0         0         3          0             4  
geller           1         1         3          1            13  
janice           0         1         0          0             8  
mike             0         0        65         26           101  
mom              7         1         7          4            68  
dad              1         2         4          5            41

In [49]:
alllines = []
alllines.append(flatten(liness1))
alllines.append(flatten(liness2))
alllines.append(flatten(liness3))
alllines.append(flatten(liness4))
alllines.append(flatten(liness5))
alllines.append(flatten(liness6))
alllines.append(flatten(liness7))
alllines.append(flatten(liness8))
alllines.append(flatten(liness9))
alllines.append(flatten(liness10))

In [50]:
saystimes = []
for line in range(len(lines)):
    saysthis = []
    for n in alllines:
        saysthis.append(sum(lines[line] in say for say in n))
    saysthis.append(sum(saysthis))    
    saystimes.append(saysthis)

pd.DataFrame(saystimes, lines,seasons)

Season 1  Season 2  Season 3  Season 4  Season 5  Season 6  \
joey           177       103       124       166       204       193   
rachel         199        83       127       106       159       194   
ross           275       126       216       180       218       250   
chandler       169       107       132       153       194       223   
monica         191        87       125        92       183       164   
phoebe         134        56        97        97       140       158   
gunther          7        10        10        12         8        11   
jack            14         8         4         9        12        15   
judy             1         1         0         2         1         2   
geller          14         4         9        10         9        13   
janice          21        11        37         9         6         0   
mike            35         0         1         0         0         0   
mom             74        44        29        46        14        20   
dad             34        51        29        20        13        30   

          Season 7  Season 8  Season 9  Season 10  All in total  
joey           217       193       187        165          1729  
rachel         127       179       131        121          1426  
ross           180       214       221        175          2055  
chandler       211       128       140        136          1593  
monica         260       108       143        107          1460  
phoebe         166       122       123         95          1188  
gunther          4         5         5          7            79  
jack             7         4        12          4            89  
judy             1         0         4          0            12  
geller          10         8        16          7           100  
janice           4         7         1          7           103  
mike             1         1       108         55           201  
mom             38        26        44         20           355  
dad             51        19        22         18           287

Gunther appears in 149 episodes - is mentioned 79 times

Jack Geller appears in 22 episodes - is mentioned 89 times but could also be called "mr. geller" or "dad"

Judy Geller appears in 20 episodes - is mentioned only 12 times but could also be called "mrs. geller" or "mom"

Janice appears in 20 episodes - is mentioned 103 times; espicially in season 1 and 3

Mike (Phoebes boyfriend/husband) appears in 18 episodes - is mentioned 201 times but mostly in season 9. In season 2 to season 8 he is only mentioned three times.


The friends mentions "mom" 1.5 times more than "dad". Espically Ross mentions "mom" (86 times) and "dad" (84 times) - maybe because he himself is a dad.